src/java.base/share/classes/java/lang/StackTraceElement.java

```diff
@@ -355,27 +355,58 @@ public boolean isNativeMethod() {
      * @revised 9
      * @see    Throwable#printStackTrace()
      */
+    @Override
     public String toString() {
-        String s = "";
-        if (!dropClassLoaderName() && classLoaderName != null &&
-                !classLoaderName.isEmpty()) {
-            s += classLoaderName + "/";
+        StringBuilder sb = new StringBuilder();
+        try {
+            appendTo(sb);
+        } catch (IOException shouldNotHappen) {
+            throw new RuntimeException("Error building StackTraceElement string", shouldNotHappen);
         }
-        if (moduleName != null && !moduleName.isEmpty()) {
-            s += moduleName;
+        return sb.toString();
+    }
 
-            if (!dropModuleVersion() && moduleVersion != null &&
-                    !moduleVersion.isEmpty()) {
-                s += "@" + moduleVersion;
+    /**
+     * Appends the {@link #toString()} representation of this stack trace element to the provided appendable.
+     * @param dest The destination appendable
+     * @throws IOException If an I/O error occurs
+     */
+    private void appendTo(Appendable dest) throws IOException {
+        boolean hasPrefix = false;
+        if (!dropClassLoaderName() && classLoaderName != null && !classLoaderName.isEmpty()) {
+            hasPrefix = true;
+            dest.append(classLoaderName).append('/');
+        }
+        if (moduleName != null && !moduleName.isEmpty()) {
+            hasPrefix = true;
+            dest.append(moduleName);
+            if (!dropModuleVersion() && moduleVersion != null && !moduleVersion.isEmpty()) {
+                dest.append('@').append(moduleVersion);
             }
         }
-        s = s.isEmpty() ? declaringClass : s + "/" + declaringClass;
+        if (hasPrefix) {
+            dest.append(declaringClass);
+        } else {
+            dest.append('/').append(declaringClass);
+        }
 
-        return s + "." + methodName + "(" +
-             (isNativeMethod() ? "Native Method)" :
-              (fileName != null && lineNumber >= 0 ?
-               fileName + ":" + lineNumber + ")" :
-                (fileName != null ?  ""+fileName+")" : "Unknown Source)")));
+        dest.append('.').append(methodName).append('(');
+        if (isNativeMethod()) {
+            dest.append("Native Method)");
+        } else {
+            if (fileName != null && lineNumber >= 0) {
+                dest.append(fileName)
+                        .append(':')
+                        .append(Integer.toString(lineNumber))
```

Comment:
I think `Integer.toString(lineNumber)` is redundant here, you can pass `int` directly

Reply:
`Appendable` does not currently have an `append(int)` method on the interface, though `StringBuilder` does. We could add it to `Appendable` via default method if so desired:

```
default Appendable append(int value) {
    return append(Integer.toString(value));
}
```

Reply:
I think in this case it's better to specify `StringBuilder` instead of `Appendable`, because the method is private and you'd hardly ever pass there anything different from SB.

Reply:
If we go with the proposal above to make `appendTo(Appendable)` package visible so that `Throwable.printStackTrace()` can benefit from this as well, we would need to keep this as Appendable, otherwise I agree we should just inline this all into toString

Reply:
`Throwable.printStackTrace()` will use `StringBuilder` as well, won't it?

Reply:
I don't think so, it appends directly do something like System.err (default) or another PrintStream/PrintWriter without fully holding the string value on heap. If it supported overloads for Appendable/StringBuilder, that would be fantastic, but that would expand the scope of this PR. Is that something you would support?

Reply:
I think we could create another PR adding suggested 
```java
default Appendable append(int value) {
    return append(Integer.toString(value));
}
```
to `Appendable`